File: TableIndexViewOC/TableIndexViewOC/TableIndexView.m
//
//  TableIndexView.m
//  SNYifubao
//
//  Created by GK on 2017/9/30.
//  Copyright © 2017年 Suning. All rights reserved.
//

#import "TableIndexView.h"

@implementation TableIndexView

- (void)setup {
    
    [self setExclusiveTouch:YES];
    self.translatesAutoresizingMaskIntoConstraints = NO;
    
    NSMutableDictionary *views = [[NSMutableDictionary alloc] init];
    NSString *verticalLayoutString = @"V:|";
    
    for (int i = 0; i < self.indexs.count; i++) {
        UILabel *label = [[UILabel alloc] initWithFrame: CGRectMake(0, i * 10, 20, 20)];
        label.text = self.indexs[i];
        label.font = [UIFont systemFontOfSize:12];
        label.textAlignment = NSTextAlignmentCenter;
        label.translatesAutoresizingMaskIntoConstraints = false;
        
        [self addSubview:label];
        NSString *indexString = [NSString stringWithFormat:@"label%d",i];
        [views setObject:label forKey:indexString];
        
        if (i == 0) {
          verticalLayoutString =  [verticalLayoutString stringByAppendingString:[NSString stringWithFormat:@"[%@]",indexString]];;
        }else {
          verticalLayoutString =  [verticalLayoutString stringByAppendingString:[NSString stringWithFormat:@"[%@(==label0)]",indexString]];
        }
        [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:[NSString stringWithFormat:@"H:|[%@]|",indexString] options:NSLayoutFormatAlignAllCenterY metrics:nil views:views]];
    }
     verticalLayoutString = [verticalLayoutString stringByAppendingString:@"|"];
     [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:verticalLayoutString options:NSLayoutFormatAlignAllCenterX metrics:nil views:views]];
    
    
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(indexViewWasDragged:)];
    UIPanGestureRecognizer *pan = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(indexViewWasDragged:)];
    [self addGestureRecognizer:tap];
    [self addGestureRecognizer:pan];
}

- (void)indexViewWasDragged:(UIGestureRecognizer *)gesture {
    CGPoint point = [gesture locationInView:self];
    NSInteger index = MAX(MIN(point.y / self.frame.size.height * self.indexs.count , self.indexs.count - 1), 0);
    CGFloat percentInSection = MAX(point.y / self.frame.size.height * self.indexs.count - index, 0);
    
    [self scrollToIndex:index percentInSection:percentInSection];
}
- (void)scrollToIndex:(NSInteger) index  percentInSection: (CGFloat)percentInSection {
    NSInteger section = index;
    NSInteger rows = [self.tableView numberOfRowsInSection:section];
    NSInteger row = rows * percentInSection;
    NSInteger numberOfSectionsInTable = [self.tableView numberOfSections];
    
//    if (row == 0 && section < numberOfSectionsInTable - 1) {
//        return ;
//    }
    if (rows != 0 && row < rows && section < numberOfSectionsInTable) {
            NSIndexPath *indexPath = [NSIndexPath indexPathForRow:row inSection:section];
            [self.tableView scrollToRowAtIndexPath:indexPath atScrollPosition:UITableViewScrollPositionTop animated:false];
    }

}
@end

File: TableIndexView/TableIndexView/TableIndexView.swift
//
//  TableIndexView.swift
//  SampleContactList
//
//  Created by GK on 2017/9/30.
//  Copyright © 2017年 Stephen Lindauer. All rights reserved.
//

import Foundation
import UIKit

class TableIndexView: UIView {
    
    var indexes: [String]!
    var tableView: UITableView!
    
    func setup() {
        
        self.isExclusiveTouch  = true
        translatesAutoresizingMaskIntoConstraints = false
        
        var views = [String:UILabel]()
        var verticalLayoutString = "V:|"
        
        for i in 0..<indexes.count {
            let label = UILabel(frame: CGRect(x: 0, y: 0, width: 20, height: 20))
            label.text = indexes[i]
            label.font = UIFont.systemFont(ofSize: 12)
            label.textAlignment = .center
            label.translatesAutoresizingMaskIntoConstraints = false
            
            addSubview(label)
            views["label\(i)"] = label
            
            if i == 0 {
                verticalLayoutString += "[label\(i)]"
            }
            else {
                verticalLayoutString += "[label\(i)(==label0)]"
            }
            
            addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|[label\(i)]|", options: NSLayoutFormatOptions.alignAllCenterY, metrics: [:], views: views))
        }
        
        verticalLayoutString += "|"
        addConstraints(NSLayoutConstraint.constraints(withVisualFormat: verticalLayoutString, options: NSLayoutFormatOptions.alignAllCenterX, metrics: [:], views: views))
        
        let panGestureRecognizer = UIPanGestureRecognizer(target: self, action: #selector(indexViewWasDragged))
        addGestureRecognizer(panGestureRecognizer)
        
        let tapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(indexViewWasDragged))
        addGestureRecognizer(tapGestureRecognizer)
    }
    
    @objc func indexViewWasDragged(_ gesture: UIPanGestureRecognizer) {
        let point = gesture.location(in: self)
        let index = max(min(Int(point.y / frame.height * CGFloat(indexes.count)), indexes.count-1), 0)
        let percentInSection = max(point.y / frame.height * CGFloat(indexes.count) - CGFloat(index), 0)
        scrollToIndex(index, percentInSection: percentInSection)
    }
    func scrollToIndex(_ index: Int, percentInSection: CGFloat) {
        var section = index
        var rows = self.tableView.dataSource!.tableView(tableView, numberOfRowsInSection: section)
        var row = Int(CGFloat(rows) * percentInSection)
        let numberOfSectionsInTable = tableView.dataSource!.numberOfSections!(in: tableView)
        
        if (rows == 0 && section < numberOfSectionsInTable-1) {
            return;
        }
        
//        while (rows == 0 && section < numberOfSectionsInTable-1) {
//            section += 1
//            rows = self.tableView.dataSource!.tableView(tableView, numberOfRowsInSection: section)
//            row = 0
//        }
//
//        if (rows != 0) {
//            let indexPath = IndexPath(row: row, section: section)
//            tableView.scrollToRow(at: indexPath, at: .top, animated: false)
//        }

        
        if (rows != 0) {
           let tmpIndex  = self.tableView.numberOfRows(inSection: section)
            if (row <= tmpIndex && row != 0 ) {
                let indexPath = IndexPath(row: row, section: section)
                tableView.scrollToRow(at: indexPath, at: .top, animated: false)
            }
          
        }
    }
}
